File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
       Stack<TreeNode> s = new Stack<>();
        List<Integer> res = new ArrayList<>();
        TreeNode currRoot = root;

        while(currRoot != null || !s.isEmpty()){
            if(currRoot != null){
                s.push(currRoot);
                currRoot = currRoot.left;
            }else{
                TreeNode right = s.peek().right;
                if(right == null){
                    right = s.pop();
                    res.add(right.val);
                    while(!s.isEmpty() && right == s.peek().right){
                        right = s.pop();
                        res.add(right.val);
                    }
                }else{
                    currRoot = right;
                }
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        left = self.isValidBST(root.left)
        right = self.isValidBST(root.right)

        lc = root.left.val < root.val if root.left else left
        rc = root.right.val > root.val if root.right else right

        return lc and rc